fem/bilininteg_mass_ea.cpp

```diff
@@ -280,4 +280,17 @@ void MassIntegrator::AssembleEA(const FiniteElementSpace &fes,
    MFEM_ABORT("Unknown kernel.");
 }
 
+void MassIntegrator::AssembleEA(const FiniteElementSpace &fes,
+                                DenseTensor &ea_tensor,
+                                const bool add)
+{
+
+   Vector ea_vec;
+   mfem::Swap(ea_vec.GetMemory(), ea_tensor.GetMemory());
```

Comment:
`ea_vec.Size()` will be `0` here, do you want to set it to `ea_tensor.TotalSize()` or does `AssembleEA` call `SetSize()`?

Reply:
Yes, that is a good catch. I'll fix that.

Reply:
Since `AssembleEA()` with `ea_vec` assumes that `ea_vec` is allocated outside, it is probably good to make the same assumption about `ea_tensor`.

To set the `Vector::size` at the same time as `Vector::data` you can use `Vector::NewMemoryAndSize` with `own_mem = false`. This will make the `ea_vec.data` an alias to `ea_tensor.tdata`. Since `ea_vec.data` is an alias, it should be sync'd with the base, `ea_tensor.tdata`, just in case the data was copied/moved. In this case, the sync can be done with `ea_tensor.GetMemory().Sync(ea_vec.GetMemory())`.

Alternatively, one can use `ea_vec.NewMemoryAndSize` with `own_mem = true`. Then after assembly into `ea_vec`, copy the `Memory` back to `ea_tensor` with `ea_tensor.GetMemory() = ea_vec.GetMemory()`. And finally, prevent `ea_vec` from deleting its `Memory` by calling `ea_vec.GetMemory().Reset()`.

Reply:
By the way, this implementation seems general and can be moved to the base class, as far as I can tell.

Reply:
The one detail about moving this to the base class is that the layout for the ea data is row major while everything else in MFEM is in column major. My feeling is this might lead to unexpected behavior in user code. Since the mass matrix is symmetric we can get away with the layout.

Reply:
It seems like there are different experiences with regard to the row-major vs column-major storage for the element matrices. If I remember correctly, @YohannDudouit said that he had concerns about the mat-vec performance when using column-major storage. Maybe it is best to have a separate discussion on the topic and run some benchmarks to help us make a decision.

Reply:
I had concerns, but I don't remember if I verified experimentally. It shouldn't be hard to add a `layout` template that can either be row-major or column-major, the implementation is fairly simple. I think each data format has its own strengths and weaknesses, one being more suited for `Mult` the other one for `MultTranspose`.

Reply:
@YohannDudouit, is the layout of the EA data documented somewhere? Can we add a note about it e.g. in the comments for the `AssembleEA` method?